django/db/models/deletion.py

```diff
@@ -306,7 +306,7 @@ def collect(
             return
 
         if keep_parents:
-            parents = set(model._meta.get_parent_list())
+            parents = set(model._meta.all_parents)
```

Comment:
Now that `all_parents` is a cached property, I think this part of the code can be removed and later just do:

```python
            # Preserve parent reverse relationships if keep_parents=True.
            if keep_parents and related.model in model._meta.all_parents:
                continue
```
More so because the highlighted `if` branch is defining a `parents` variable that will produce a `NameError` if `keep_parents` is False (and that is hard to spot in the code, and also a code smell IMHO).

Reply:
Nice spot. The suggestion *will* change the lookup behaviour from O(1) to O(n), but I think that is likely okay given the typical shallowness of model inheritance where 10 layers would be excessive.